compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -8999,8 +8999,7 @@ private BType checkTupleIndexBasedAccess(BLangIndexBasedAccess accessExpr, BTupl
                 return checkTupleFieldType(tuple, getConstIndex(indexExpr).intValue());
             }
 
-            BTupleType tupleExpr = (BTupleType) accessExpr.expr.getBType();
-            LinkedHashSet<BType> tupleTypes = collectTupleFieldTypes(tupleExpr, new LinkedHashSet<>());
+            LinkedHashSet<BType> tupleTypes = collectTupleFieldTypes(tuple, new LinkedHashSet<>());
```

Comment:
Can there be a scenario where `tuple != accessExpr.expr.getBType()`?

Reply:
I will chek.

Reply:
For this type
```
type IntUnionStringTuple [int...]|[string...];
```

This `checkTupleIndexBasedAccess` method will call twice.

In both iterations `accessExpr.expr.getBType() = IntUnionStringTuple` which is a `BTypeReferenceType`

But `tuple = [int...]` in first iteration and `tuple = [string...]` in second iteration.

We have to use `tuple` as the first argument for the `collectTupleFieldTypes` to resolve the issue.

Reply:
Add such test cases. If both of the member types in the union reaches this point; then fine